net/dns/manager.go

```diff
@@ -476,13 +484,15 @@ func (s *dnsTCPSession) handleQuery(q []byte) {
 		return
 	}
 
+	// See note in handleWrites (above) regarding this select{}
 	select {
 	case <-s.ctx.Done():
 	case s.responses <- resp:
 	}
 }
 
 func (s *dnsTCPSession) handleReads() {
+	defer s.conn.Close()
```

Comment:
is this an unrelated fix?

Reply:
Gotcha. Do we need both conn.Close boths? Looks like it's done in both read and write exit paths?

Reply:
I think so? The same issue occurs there; if we're trying to write to a full `conn` (or a `net.Pipe` in tests), then `handleReads` can read something, bail due the size limit, and any in-progress writes can block if the client isn't reading. I guess it will only block for `idleTimeoutTCP` time, though, since we call `SetWriteDeadline`, but I'd prefer not to rely on that :smiley: